File: src/services/routeGeometry.ts
import type {
  ImportedRouteGeometry,
  ResampledRouteGeometry,
  RouteCoordinate,
} from "../types/route";

export const DEFAULT_ROUTE_SAMPLE_SPACING_M = 40;
export const MAX_ROUTE_SAMPLES = 6000;
export const MAX_GPX_SOURCE_POINTS = 100000;
export const MAX_GPX_FILE_BYTES = 15 * 1024 * 1024;
const EARTH_RADIUS_M = 6371008.8;
const CONTIGUOUS_SEGMENT_GAP_M = 500;

interface GpxElementLike {
  getAttribute(name: string): string | null;
  getElementsByTagNameNS(namespace: string, localName: string): ArrayLike<GpxElementLike>;
  textContent: string | null;
}

interface GpxDocumentLike {
  getElementsByTagNameNS(namespace: string, localName: string): ArrayLike<GpxElementLike>;
  querySelector(selector: string): GpxElementLike | null;
}

function radians(value: number): number {
  return (value * Math.PI) / 180;
}

export function shortestLongitudeDelta(from: number, to: number): number {
  return ((to - from + 540) % 360) - 180;
}

export function normaliseLongitude(longitude: number): number {
  return ((longitude + 540) % 360) - 180;
}

export function unwrapRouteCoordinates(
  coordinates: RouteCoordinate[]
): RouteCoordinate[] {
  if (coordinates.length === 0) return [];
  const unwrapped: RouteCoordinate[] = [
    { ...coordinates[0], longitude: normaliseLongitude(coordinates[0].longitude) },
  ];
  for (let index = 1; index < coordinates.length; index++) {
    const previous = unwrapped[index - 1];
    unwrapped.push({
      latitude: coordinates[index].latitude,
      longitude:
        previous.longitude +
        shortestLongitudeDelta(previous.longitude, coordinates[index].longitude),
    });
  }
  return unwrapped;
}

export function geodesicDistanceM(
  first: RouteCoordinate,
  second: RouteCoordinate
): number {
  const latitude1 = radians(first.latitude);
  const latitude2 = radians(second.latitude);
  const latitudeDelta = latitude2 - latitude1;
  const longitudeDelta = radians(
    shortestLongitudeDelta(first.longitude, second.longitude)
  );
  const haversine =
    Math.sin(latitudeDelta / 2) ** 2 +
    Math.cos(latitude1) *
      Math.cos(latitude2) *
      Math.sin(longitudeDelta / 2) ** 2;
  return 2 * EARTH_RADIUS_M * Math.asin(Math.min(1, Math.sqrt(haversine)));
}

function parseCoordinate(element: GpxElementLike): RouteCoordinate | null {
  const latitude = Number(element.getAttribute("lat"));
  const longitude = Number(element.getAttribute("lon"));
  if (
    !Number.isFinite(latitude) ||
    !Number.isFinite(longitude) ||
    latitude < -90 ||
    latitude > 90 ||
    longitude < -180 ||
    longitude > 180
  ) {
    return null;
  }
  return { latitude, longitude };
}

function coordinatesFrom(
  element: GpxElementLike,
  pointName: "trkpt" | "rtept"
): RouteCoordinate[] {
  return Array.from(element.getElementsByTagNameNS("*", pointName))
    .map(parseCoordinate)
    .filter((point): point is RouteCoordinate => point !== null);
}

function routeLength(points: RouteCoordinate[]): number {
  return points.slice(1).reduce(
    (distance, point, index) => distance + geodesicDistanceM(points[index], point),
    0
  );
}

export function chooseGpxRouteCoordinates(
  trackSegments: RouteCoordinate[][],
  routeCandidates: RouteCoordinate[][]
): { coordinates: RouteCoordinate[]; sourceSegmentCount: number } {
  const usableSegments = trackSegments.filter((segment) => segment.length >= 2);
  const chains: RouteCoordinate[][] = [];
  for (const segment of usableSegments) {
    const current = chains[chains.length - 1];
    if (
      current &&
      geodesicDistanceM(current[current.length - 1], segment[0]) <=
        CONTIGUOUS_SEGMENT_GAP_M
    ) {
      current.push(
        ...(geodesicDistanceM(current[current.length - 1], segment[0]) < 0.5
          ? segment.slice(1)
          : segment)
      );
    } else {
      chains.push([...segment]);
    }
  }
  const candidates = [
    ...chains,
    ...routeCandidates.filter((candidate) => candidate.length >= 2),
  ];
  if (candidates.length === 0) {
    throw new Error("The GPX file does not contain a usable track or route.");
  }
  const coordinates = candidates.reduce((longest, candidate) =>
    routeLength(candidate) > routeLength(longest) ? candidate : longest
  );
  if (coordinates.length > MAX_GPX_SOURCE_POINTS) {
    throw new Error(
      `The GPX contains more than ${MAX_GPX_SOURCE_POINTS.toLocaleString()} usable points.`
    );
  }
  return {
    coordinates,
    sourceSegmentCount: usableSegments.length || routeCandidates.length,
  };
}

export function parseGpxDocument(
  document: GpxDocumentLike,
  fallbackName = "Imported route"
): ImportedRouteGeometry {
  if (document.querySelector("parsererror")) {
    throw new Error("The selected file is not valid GPX/XML.");
  }
  const trackSegments = Array.from(
    document.getElementsByTagNameNS("*", "trkseg")
  ).map((segment) => coordinatesFrom(segment, "trkpt"));
  const routeCandidates = Array.from(
    document.getElementsByTagNameNS("*", "rte")
  ).map((route) => coordinatesFrom(route, "rtept"));
  const selected = chooseGpxRouteCoordinates(trackSegments, routeCandidates);
  const nameElement = Array.from(
    document.getElementsByTagNameNS("*", "name")
  )[0];
  const name = nameElement?.textContent?.trim() || fallbackName;
  return {
    id: `route-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
    name,
    coordinates: selected.coordinates,
    sourcePointCount: selected.coordinates.length,
    sourceSegmentCount: selected.sourceSegmentCount,
  };
}

export function parseGpxText(
  text: string,
  fallbackName = "Imported route"
): ImportedRouteGeometry {
  const document = new DOMParser().parseFromString(text, "application/xml");
  return parseGpxDocument(document, fallbackName);
}

function interpolateCoordinate(
  first: RouteCoordinate,
  second: RouteCoordinate,
  fraction: number
): RouteCoordinate {
  return {
    latitude: first.latitude + (second.latitude - first.latitude) * fraction,
    longitude: normaliseLongitude(
      first.longitude + shortestLongitudeDelta(first.longitude, second.longitude) * fraction
    ),
  };
}

export function resampleRouteGeometry(
  route: ImportedRouteGeometry,
  requestedSpacingM = DEFAULT_ROUTE_SAMPLE_SPACING_M
): ResampledRouteGeometry {
  if (route.coordinates.length < 2) {
    throw new Error("A route needs at least two usable coordinates.");
  }
  const sourceDistances = [0];
  for (let index = 1; index < route.coordinates.length; index++) {
    sourceDistances.push(
      sourceDistances[index - 1] +
        geodesicDistanceM(route.coordinates[index - 1], route.coordinates[index])
    );
  }
  const totalDistanceM = sourceDistances[sourceDistances.length - 1];
  if (!Number.isFinite(totalDistanceM) || totalDistanceM < 10) {
    throw new Error("The GPX route is too short for journey analysis.");
  }
  const spacingM = Math.max(
    requestedSpacingM,
    totalDistanceM / Math.max(1, MAX_ROUTE_SAMPLES - 1)
  );
  const targetDistances: number[] = [];
  for (let distance = 0; distance < totalDistanceM; distance += spacingM) {
    targetDistances.push(distance);
  }
  targetDistances.push(totalDistanceM);
  const coordinates: RouteCoordinate[] = [];
  let sourceIndex = 1;
  for (const targetDistance of targetDistances) {
    while (
      sourceIndex < sourceDistances.length - 1 &&
      sourceDistances[sourceIndex] < targetDistance
    ) {
      sourceIndex += 1;
    }
    const beforeDistance = sourceDistances[sourceIndex - 1];
    const afterDistance = sourceDistances[sourceIndex];
    const fraction =
      afterDistance === beforeDistance
        ? 0
        : (targetDistance - beforeDistance) / (afterDistance - beforeDistance);
    coordinates.push(
      interpolateCoordinate(
        route.coordinates[sourceIndex - 1],
        route.coordinates[sourceIndex],
        Math.max(0, Math.min(1, fraction))
      )
    );
  }
  return {
    id: route.id,
    name: route.name,
    coordinates,
    cumulativeDistancesM: targetDistances,
    totalDistanceM,
    spacingM,
    sourcePointCount: route.sourcePointCount,
  };
}

export function getRouteBounds(coordinates: RouteCoordinate[]): {
  west: number;
  south: number;
  east: number;
  north: number;
} {
  const latitudes = coordinates.map((point) => point.latitude);
  const wrapped = coordinates
    .map((point) => ((point.longitude % 360) + 360) % 360)
    .sort((a, b) => a - b);
  let largestGap = -1;
  let gapIndex = 0;
  for (let index = 0; index < wrapped.length; index++) {
    const next = index === wrapped.length - 1 ? wrapped[0] + 360 : wrapped[index + 1];
    const gap = next - wrapped[index];
    if (gap > largestGap) {
      largestGap = gap;
      gapIndex = index;
    }
  }
  const westWrapped = wrapped[(gapIndex + 1) % wrapped.length];
  const eastWrapped = wrapped[gapIndex] + (gapIndex === wrapped.length - 1 ? 0 : 360);
  const west = westWrapped > 180 ? westWrapped - 360 : westWrapped;
  let east = eastWrapped > 180 ? eastWrapped - 360 : eastWrapped;
  while (east < west) east += 360;
  return {
    west,
    east,
    south: Math.min(...latitudes),
    north: Math.max(...latitudes),
  };
}

File: scripts/terrain_research/route_geometry.py
from __future__ import annotations

import math
import xml.etree.ElementTree as ET
from dataclasses import dataclass
from pathlib import Path
from typing import Sequence


EARTH_RADIUS_M = 6_371_008.8
MAX_ROUTE_SAMPLES = 100_000
CONTIGUOUS_SEGMENT_GAP_M = 500.0


@dataclass(slots=True, frozen=True)
class RouteCoordinate:
    latitude: float
    longitude: float


@dataclass(slots=True, frozen=True)
class RouteGeometry:
    name: str
    coordinates: tuple[RouteCoordinate, ...]
    cumulative_distances_m: tuple[float, ...]
    total_distance_m: float


def shortest_longitude_delta(start: float, end: float) -> float:
    return ((end - start + 540.0) % 360.0) - 180.0


def geodesic_distance_m(first: RouteCoordinate, second: RouteCoordinate) -> float:
    latitude_1 = math.radians(first.latitude)
    latitude_2 = math.radians(second.latitude)
    latitude_delta = latitude_2 - latitude_1
    longitude_delta = math.radians(
        shortest_longitude_delta(first.longitude, second.longitude)
    )
    haversine = (
        math.sin(latitude_delta / 2.0) ** 2
        + math.cos(latitude_1)
        * math.cos(latitude_2)
        * math.sin(longitude_delta / 2.0) ** 2
    )
    return 2.0 * EARTH_RADIUS_M * math.asin(min(1.0, math.sqrt(haversine)))


def _valid_coordinate(element: ET.Element) -> RouteCoordinate | None:
    try:
        latitude = float(element.attrib["lat"])
        longitude = float(element.attrib["lon"])
    except (KeyError, TypeError, ValueError):
        return None
    if not (-90.0 <= latitude <= 90.0 and -180.0 <= longitude <= 180.0):
        return None
    return RouteCoordinate(latitude, longitude)


def _local_name(tag: str) -> str:
    return tag.rsplit("}", 1)[-1]


def _length(coordinates: Sequence[RouteCoordinate]) -> float:
    return sum(
        geodesic_distance_m(first, second)
        for first, second in zip(coordinates, coordinates[1:])
    )


def _choose_coordinates(root: ET.Element) -> list[RouteCoordinate]:
    segments: list[list[RouteCoordinate]] = []
    routes: list[list[RouteCoordinate]] = []
    for element in root.iter():
        name = _local_name(element.tag)
        if name not in {"trkseg", "rte"}:
            continue
        point_name = "trkpt" if name == "trkseg" else "rtept"
        points = [
            point
            for child in element.iter()
            if _local_name(child.tag) == point_name
            for point in [_valid_coordinate(child)]
            if point is not None
        ]
        if len(points) >= 2:
            (segments if name == "trkseg" else routes).append(points)

    chains: list[list[RouteCoordinate]] = []
    for segment in segments:
        if (
            chains
            and geodesic_distance_m(chains[-1][-1], segment[0])
            <= CONTIGUOUS_SEGMENT_GAP_M
        ):
            chains[-1].extend(
                segment[1:]
                if geodesic_distance_m(chains[-1][-1], segment[0]) < 0.5
                else segment
            )
        else:
            chains.append(list(segment))
    candidates = chains + routes
    if not candidates:
        raise ValueError("GPX does not contain a usable track or route")
    return max(candidates, key=_length)


def parse_gpx_geometry(path: Path) -> RouteGeometry:
    try:
        root = ET.parse(path).getroot()
    except (ET.ParseError, OSError) as error:
        raise ValueError(f"Unable to parse GPX geometry: {path.name}") from error
    coordinates = _choose_coordinates(root)
    distances = [0.0]
    for first, second in zip(coordinates, coordinates[1:]):
        distances.append(distances[-1] + geodesic_distance_m(first, second))
    if distances[-1] < 10.0:
        raise ValueError("Route is too short for terrain analysis")
    name = path.stem
    for element in root.iter():
        if _local_name(element.tag) == "name" and element.text and element.text.strip():
            name = element.text.strip()
            break
    return RouteGeometry(name, tuple(coordinates), tuple(distances), distances[-1])


def _interpolate(
    first: RouteCoordinate, second: RouteCoordinate, fraction: float
) -> RouteCoordinate:
    longitude = first.longitude + shortest_longitude_delta(
        first.longitude, second.longitude
    ) * fraction
    longitude = ((longitude + 540.0) % 360.0) - 180.0
    return RouteCoordinate(
        first.latitude + (second.latitude - first.latitude) * fraction,
        longitude,
    )


def resample_route(route: RouteGeometry, spacing_m: float) -> RouteGeometry:
    if not math.isfinite(spacing_m) or spacing_m <= 0:
        raise ValueError("Route sampling interval must be positive")
    target_count = math.ceil(route.total_distance_m / spacing_m) + 1
    if target_count > MAX_ROUTE_SAMPLES:
        raise ValueError(
            f"Route would exceed the explicit {MAX_ROUTE_SAMPLES:,}-sample limit"
        )
    targets = [index * spacing_m for index in range(target_count - 1)]
    if not targets or route.total_distance_m - targets[-1] > 0.01:
        targets.append(route.total_distance_m)
    else:
        targets[-1] = route.total_distance_m
    result: list[RouteCoordinate] = []
    source_index = 1
    for target in targets:
        while (
            source_index < len(route.cumulative_distances_m) - 1
            and route.cumulative_distances_m[source_index] < target
        ):
            source_index += 1
        before = route.cumulative_distances_m[source_index - 1]
        after = route.cumulative_distances_m[source_index]
        fraction = 0.0 if after == before else (target - before) / (after - before)
        result.append(
            _interpolate(
                route.coordinates[source_index - 1],
                route.coordinates[source_index],
                min(1.0, max(0.0, fraction)),
            )
        )
    return RouteGeometry(route.name, tuple(result), tuple(targets), route.total_distance_m)
